### PR TITLE
support retry and dead sets from sidekiq

### PR DESCRIPTION
**Pull request by xldenis**
Add support for the `RetrySet` and `DeadSet` queues in sidekiq.

### CODE REVIEW

File: lib/tasks/sidekiq.rake [hunk 1]
```diff
@@ -43,14 +43,28 @@ namespace :sidekiq do
...
         source: SidekiqDelayedQueue.new(redis: source_redis),
         dest:   SidekiqDelayedQueue.new(redis: dest_redis),
         logger: logger,
         progress: progress
-        )
+      )
+
+      retry_queue = SidekiqMover.new(
+        source: SidekiqDelayedQueue.new(key: 'retry', redis: source_redis),
+        dest: SidekiqDelayedQueue.new(key: 'retry', redis: dest_redis),
```
Comment (fbogsany) [inline lib/tasks/sidekiq.rake]: Minor niggle: indentation of `dest:` arg here and below doesn't match other calls.

**Comment by fbogsany** [comment]
Can you clarify what these queues are? E.g. references to the corresponding Sidekiq code?

**Comment by xldenis** [comment]
:+1:

I've added links to the source definitions of the Retry and Dead sets. They are queues used by sidekiq to store jobs that need to retry and jobs that have used up all their retries (sidekiq doesn't just delete them).

**Review by shivnagarajan — APPROVED** [review]
👍 for the links on `retry_queue` and `dead_queue`

Could we put these links into the Readme under Sidekiq?

**Comment by xldenis** [comment]
> Could we put these links into the Readme under Sidekiq?

Do you think this actually helps in any way? Sidekiq users will already be familiar with RetrySet and DeadSet and don't need to know the specific implementation of them.

File: lib/tasks/sidekiq.rake [hunk 2]
```diff
@@ -43,14 +43,29 @@ namespace :sidekiq do
...
 
       delayed_queue = SidekiqMover.new(
         source: SidekiqDelayedQueue.new(redis: source_redis),
-        dest:   SidekiqDelayedQueue.new(redis: dest_redis),
+        dest: SidekiqDelayedQueue.new(redis: dest_redis),
         logger: logger,
         progress: progress
-        )
+      )
+      # https://github.com/mperham/sidekiq/blob/63ee43353bd3b753beb0233f64865e658abeb1c3/lib/sidekiq/api.rb#L641
```
Comment (fbogsany) [inline lib/tasks/sidekiq.rake]: Add a blank line before this.

Comment (xldenis) [reply lib/tasks/sidekiq.rake]: :+1:

**Comment by xldenis** [comment]
I've rebased these changes on latest master, I'm writing test cases for it now.

File: lib/tasks/sidekiq.rake (hunk 2 above)
Comment (fbogsany) [reply lib/tasks/sidekiq.rake]: https://github.com/mperham/sidekiq/wiki/API#retries might be a better link?

File: lib/tasks/sidekiq.rake [hunk 3]
```diff
@@ -43,14 +43,29 @@ namespace :sidekiq do
...
+      retry_queue = SidekiqMover.new(
+        source: SidekiqDelayedQueue.new(key: 'retry', redis: source_redis),
+        dest: SidekiqDelayedQueue.new(key: 'retry', redis: dest_redis),
+        logger: logger,
+        progress: progress,
+      )
+
+      # https://github.com/mperham/sidekiq/blob/63ee43353bd3b753beb0233f64865e658abeb1c3/lib/sidekiq/api.rb#L656
```
Comment (fbogsany) [inline lib/tasks/sidekiq.rake:63]: https://github.com/mperham/sidekiq/wiki/API#dead might be a better link?

File: lib/tasks/sidekiq.rake (hunk 2 above)
Comment (xldenis) [reply lib/tasks/sidekiq.rake]: Not really, it doesn't actually mention the name of the redis queue which is what matters here, no?

**Comment by shivnagarajan** [comment]
>> Could we put these links into the Readme under Sidekiq?

> Do you think this actually helps in any way? Sidekiq users will already be familiar with RetrySet and DeadSet and don't need to know the specific implementation of them.

That's fair. You can leave it right here.

File: lib/tasks/sidekiq.rake (hunk 3 above)
Comment (xldenis) [reply lib/tasks/sidekiq.rake:63]: I don't know if you saw my comment on the retry thing before it became 'outdated' but basically, I feel like what matters here is the _implementation_. We want to know that it's `retry` and not `retries`.

**Comment by xldenis** [comment]
Bump for reviews! I've added a test case for dead and retry transfers.

File: test/unit/sidekiq_mover_test.rb [hunk 4]
```diff
@@ -67,14 +67,58 @@ def test_releases_locks_in_source_dc
     assert_equal [1, 2], JobHelper.job_result_queue.map {|arg| arg[0]['num']}
   end
 
+  def test_can_move_retry_queue
+    CrashingJob.perform_async(num: 1)
```
Comment (shivnagarajan) [inline test/unit/sidekiq_mover_test.rb]: can we test with > 1 . to ensure _all_ jobs are moved

File: test/unit/sidekiq_mover_test.rb [hunk 5]
```diff
@@ -67,14 +67,58 @@ def test_releases_locks_in_source_dc
...
+
+    begin
+      work_off_jobs(@source)
+    rescue CrashingJob::Error
+    end
+
+    assert_equal 1, @source.zcard('retry')
+
+    sidekiq_delayed_mover('retry')
+    assert_equal 1, @dest.zcard('retry')
```
Comment (shivnagarajan) [inline test/unit/sidekiq_mover_test.rb]: if the item is not removed from `@source` can we assert that the number in `@dest` == number in `@source`

Comment (xldenis) [reply test/unit/sidekiq_mover_test.rb]: It should be removed from the source though.

Comment (shivnagarajan) [reply test/unit/sidekiq_mover_test.rb]: If that's the case we can asset that source is 0 and dest has the right number

File: test/unit/sidekiq_mover_test.rb (hunk 4 above)
Comment (xldenis) [reply test/unit/sidekiq_mover_test.rb]: I've adjusted the test to push two jobs.

Comment (xldenis) [reply test/unit/sidekiq_mover_test.rb]: Tho I don't think that's important to check since it's tested more directly in other tests.

Comment (xldenis) [reply test/unit/sidekiq_mover_test.rb]: What really matters is that when we specify the queue `retry` that we actually get the retry queue.

File: lib/tasks/sidekiq.rake (hunk 3 above)
Comment (fbogsany) [reply lib/tasks/sidekiq.rake:63]: Perhaps this is naive, but wouldn’t it be safer to retrieve the queue name with `Sidekiq::RetrySet.new.name`? Then you’re guaranteed to have the right name.

**Review by shivnagarajan — APPROVED** [review]
Looks good to me, if we can use the `Sidekiq::RetrySet.new.name` and similarly for the `DeadSet` even better.

**Comment by xldenis** [comment]
Done! I've used the `DeadSet.new.name` directly.

EDIT: I'm undoing that change because creating a `RetrySet` / `DeadSet` connects to redis in an uncontrolled manner.

**Pull request merged**